File: csharp/src/Training.Patterns/Fn/FunctionalStyleObserver.cs
namespace Training.Patterns.Fn;

public class FunctionalStyleObserver
{
    internal class EventSource
    {
        private readonly List<Action<string>> observers = [];

        public void Subscribe(Action<string> observer)
        {
            observers.Add(observer);
        }

        public void Publish(string @event)
        {
            observers.ForEach(o => o(@event));
        }
    }

    public static void Run()
    {
        var source = new EventSource();
        source.Subscribe(msg => Console.WriteLine("Observer 1 received: " + msg));
        source.Subscribe(msg => Console.WriteLine("Observer 2 received: " + msg.ToUpperInvariant()));
        source.Publish("Functional Observer pattern with lambdas!");
    }
}

File: typescript/src/patterns/fn/FunctionalStyleObserver.ts
/** Java: Consumer<T>. */
type Consumer<T> = (value: T) => void;

// Java: static nested class FunctionalStyleObserver.EventSource
export class EventSource {
  private readonly observers: Consumer<string>[] = [];

  subscribe(observer: Consumer<string>): void {
    this.observers.push(observer);
  }

  publish(event: string): void {
    this.observers.forEach((o) => o(event));
  }
}

export function run(): void {
  const source = new EventSource();
  source.subscribe((msg) => console.log(`Observer 1 received: ${msg}`));
  source.subscribe((msg) => console.log(`Observer 2 received: ${msg.toUpperCase()}`));
  source.publish('Functional Observer pattern with lambdas!');
}
